service/qos/service_level_controller.hh

```diff
@@ -67,6 +68,23 @@ public:
         virtual future<> drop_service_level(sstring service_level_name) const = 0;
     };
     using service_level_distributed_data_accessor_ptr = ::shared_ptr<service_level_distributed_data_accessor>;
+
+    class subscriber_registry {
```

Comment:
Why the wrapper? It's more common to keep a vector of subscriber pointers directly, which makes the API simpler and avoid another level of indirection. Example: https://github.com/scylladb/scylla/blob/b503ec36c22856cf9e1d02ff0055767001be6f66/service/endpoint_lifecycle_subscriber.hh#L93

Reply:
I thought that a race might occur, but the more I think about it It probably can't happen since the tasks are executed in an
orderly fashion.
For completeness I will describe this (impossible) race that I had in mind. However, I am going to change the code.
1. A certain object registers for the API.
2. A parallel for each is queuing a call to this object's callback (which is not executed)
3. The object preregisters and destroys itself.
4. Then the queued task from 2 is executed on a dangling pointer.

The thing I didn't consider was that the tasks can't be executed out of order so it can't happen :slightly_smiling_face:

Reply:
removed in the new version